File: week-7/go/membersys/core/unit_of_work.go
package core

type UnitOfWork struct {
	MemberRepository  MemberRepository
	MessageRepository MessageRepository
}

func (unitOfWork *UnitOfWork) Init() {
}

type UnitOfWorkFactory interface {
	Init()
}

func InitUnitOfWork(factory UnitOfWorkFactory) {
	factory.Init()
}

File: week-7/py/member_system/repository/memory/unit_of_work.py
from member_system.core import MemberRepository
from member_system.core import MessageRepository
from member_system.core import UnitOfWork
from member_system.repository.memory.member_repository import MemoryMemberRepository
from member_system.repository.memory.message_repository import MemoryMessageRepository

class MemoryUnitOfWork(UnitOfWork):
    def _createMemberRepository(self) -> MemberRepository:
        return MemoryMemberRepository()
    
    def _createMessageRepository(self) -> MessageRepository:
        return MemoryMessageRepository(self.memberRepository)
